Create JsRuntimeOp Abstraction and Migrate To It

zicklag opened the pull request:
- Creates a `JsRuntimeOp` trait that can be used to implement custom ops
- Makes ops runtime agnostic, seriously cutting down on code duplication
- Migrates native runtime to use new op abstraction
- TODO: Migrate WASM runtime

### Related Issues

- Closes #16
- Closes #18

zicklag committed:
Create JsRuntimeOp Abstraction and Migrate To It

- Creates a `JsRuntimeOp` trait that can be used to implement custom ops
- Makes ops runtime agnostic, seriously cutting down on code duplication
- Migrates native runtime to use new op abstraction
- TODO: Migrate WASM runtime

zicklag committed:
Forbid Unsafe

We can actually do this which is surprixing for a scripting plugin! 🎉

jakobhellermann commented:
Awesome, this looks great from a quick look.

What's the reason for having the `JsRuntimeOp` in an `Arc`? Couldn't it be a `Box<dyn JsRuntimeOp>` as well? I don't see where op would be shared.

zicklag commented:
I think it could be a box. I was just cloning it out of the `JsRuntimeConfig` resource so I put it in an Arc, but since `JsRuntimeConfig` is only used once during initialization I could just move it instead of cloning it.

zicklag committed:
Automatically Garbage Collect Value Refs

jakobhellermann commented on src/runtime/native.rs:
```diff
@@ -265,6 +265,17 @@ impl JsRuntimeApi for JsRuntime {
+
+        // Run the JS event loop
+        with_world(world, &mut this.runtime, |runtime| {
+            if let Err(e) = pollster::block_on(runtime.run_event_loop(true)) {
```
According to `run_event_loop` docs, it resolves when
This future resolves when:
>    there are no more pending dynamic imports
>    there are no more pending ops
>    there are no more active inspector sessions (only if wait_for_inspector is set to true)

I don't think we need to block all bevy systems for when these things are true, right? 

We could call `poll_event_loop` ("Runs a single tick of event loop") but it wants a `&mut std::task::Context` which is usually provided by the future.
We could create the `Context` from a noop waker, which should be fine because we will poll every frame anyways. But only once... idk if that is bad.

Since running the event loop doesn't depend on the `World` we could probably even run it in parallel to the ecs schedule, but `run_event_loop` mutably borrows the `runtime`...

Not sure what the best solution is, I'm fine with leaving this as is for this PR.

zicklag replied:
> Since running the event loop doesn't depend on the `World`...

It actually does depend on world right now because the garbage collection hook needs access to the `JsValueRefs` resource. The world was just a convenient place to store the state, but we could move it.

I'm not sure if moving it helps, though, because, like you said we need mutable access to the runtime, and the runtime might be needed at any Bevy stage, because there might be scripts running at any stage.

As it stands, we don't have any async ops or dynamic imports, so I _think_ that `run_event_loop` should pretty much always finish with nothing but synchronous tasks, and we probably don't need to worry about it unless we expose any async ops. And we may not need scripts to do anything async anyway, since they're currently expected to run synchronous like a Bevy system, with async stuff hidden behind asset handles, and similar concepts.

jakobhellermann replied:
Okay, sounds good then.

jakobhellermann commented on src/lib.rs:
```diff
@@ -26,6 +45,29 @@ impl Plugin for JsScriptingPlugin {
+        app.add_system_to_stage(
+            CoreStage::First,
+            (|world: &mut World| {
+                let runtime = world.remove_non_send_resource::<JsRuntime>().unwrap();
+                runtime.frame_start(world);
+                world.insert_non_send_resource(runtime);
```
It would be useful to have `resource_scope` for nonsend resources here: https://github.com/bevyengine/bevy/issues/6037.

jakobhellermann commented on src/runtime/native.rs:
```diff
@@ -40,21 +34,101 @@ struct LoadedScriptData {
+        // Run our initialization script, inserting the op name mapping into the script before running it
+        let op_map_json = serde_json::to_string(&op_indexes).unwrap();
+        let init_script_src = &include_str!("./js/native_setup.js")
+            .replace("__OP_NAME_MAP_PLACEHOLDER__", &op_map_json);
```
This looks a bit hacky but at least it only happens once on initialization.

zicklag replied:
Yeah, we might be able to evaluate the JS and receive a function that we could pass our JSON object into to avoid the string substitution, but I feel like it wouldn't be gaining us much and it would make it harder to read.

Since we are statically including the `native_setup.js` at compile time I think it's safe enough.

jakobhellermann commented on src/runtime/js/native_setup.js:
```diff
@@ -0,0 +1,15 @@
+"use_strict";
```
```suggestion
"use strict";
```

jakobhellermann commented on src/runtime/ops/log.rs:
```diff
@@ -0,0 +1,77 @@
+            event!(target: "js_runtime", Level::ERROR, "{text}");
+        } else {
+            anyhow::bail!("Invalid log level");
+        }
```
superficial comment, but could you change this to a
```rust
match level {
  0 => ..,
  1 => ..,
  _ => anyhow::bail!(),
}
```
?

zicklag committed:
Migrate WASM Runtime to New Op Abstraction

Also moves op state out of the world and into the runtime
implementation.

zicklag commented:
I just pushed a commit that updates the WASM runtime, making this feature complete!

The update also moved the `JsValueRefs` and `JsReflectFunctions` out of the world and into a new `TypeMap` that represents a shared state for use in op implementations. This turned out to be important for garbage collection on WASM, which, unlike our native implementation, can have it's callback hook called at any time, not just at the end of the frame, where it has access to the world.

So moving the value refs into the runtime state allows us to free value refs without having to borrow the world.

---

I also moved some types into more appropriate modules.

zicklag committed:
Export Types Required For Implementing Ops

jakobhellermann commented on src/runtime/wasm.rs:
```diff
@@ -1,51 +1,21 @@
-
 #[wasm_bindgen]
 struct BevyModJsScripting {
     state: Rc<Mutex<JsRuntimeState>>,
```
I think we can use `Rc<RefCell<>>` here because we are just on one thread.

jakobhellermann replied:
Wait I missed that his is in the `wasm` module, and it seems that you chose to use `wasm_mutex` https://github.com/ncomendant/wasm_mutex.

jakobhellermann commented:
Is this ready to be merged or is there anything missing?

jakobhellermann committed:
put op_state and script_info into OpContext for easy extensibility

jakobhellermann commented:
I tried this out and it seems to be working fine so I'm gonna go ahead and merge it, if there's anything missing we can fix it in follow up PRs.
